File: infrastructure/src/main/java/com/study/platform/domain/user/infrastructure/UserRepositoryAdapter.java
package com.study.platform.domain.user.infrastructure;

import com.study.platform.domain.user.model.User;
import com.study.platform.domain.user.model.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
@RequiredArgsConstructor
public class UserRepositoryAdapter implements UserRepository {

    private final UserJpaRepository userJpaRepository;

    @Override
    public User save(User user) {
        return userJpaRepository.save(user);
    }

    @Override
    public Optional<User> findById(UUID id) {
        return userJpaRepository.findById(id);
    }

    @Override
    public Optional<User> findByKakaoId(String kakaoId) {
        return userJpaRepository.findByKakaoId(kakaoId);
    }

    @Override
    public Optional<User> findByNickname(String nickname) {
        return userJpaRepository.findByNickname(nickname);
    }

    @Override
    public List<User> findAllByNicknameIn(Collection<String> nicknames) {
        return userJpaRepository.findAllByNicknameIn(nicknames);
    }
}

File: infrastructure/src/main/kotlin/com/study/platform/domain/user/infrastructure/UserRepositoryAdapter.kt
package com.study.platform.domain.user.infrastructure

import com.study.platform.domain.user.model.User
import com.study.platform.domain.user.model.UserRepository
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Repository
class UserRepositoryAdapter(
    private val userJpaRepository: UserJpaRepository
) : UserRepository {

    override fun save(user: User): User =
        userJpaRepository.save(user)

    // 신규 유저 생성은 별도 트랜잭션으로 분리해, unique 제약 위반 시 flush 실패가
    // 호출자의 트랜잭션(예: AuthService.kakaoLogin)의 영속성 컨텍스트를 오염시키지 않도록 한다.
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    override fun saveNew(user: User): User =
        userJpaRepository.saveAndFlush(user)

    override fun findById(id: UUID): User? =
        userJpaRepository.findById(id).orElse(null)

    override fun findByKakaoId(kakaoId: String): User? =
        userJpaRepository.findByKakaoId(kakaoId)

    override fun findByNickname(nickname: String): User? =
        userJpaRepository.findByNickname(nickname)

    override fun findAllByNicknameIn(nicknames: Collection<String>): List<User> =
        userJpaRepository.findAllByNicknameIn(nicknames)
}
